### PR TITLE
Add the Concat Slice Elimination transform, fix constant_folding transform

### PR DESCRIPTION
**Pull request by ashbhandare**
This transform removes the redundant Concat + Slice pattern in the application of the attention bias, found in the BART model.

Before transform:
<pre>
q_bias              -- >Slice [0, q] ---> Add_q  

       \            | 

k_bias-- Concat --> -->Slice [q, q+k]---> Add_k

       /            | 

v_bias              -->Slice [q+k, :]---> Add_v
</pre>
After transform:
q_bias ---> Add_q        
k_bias ---> Add_k
v_bias ---> Add_v

This PR also fixes removal of dangling input nodes to the nodes removed using the constant folding transform.
Before transform:.
<pre>
input_x -- >transpose ->Shape ...--> Concat  
                                                          \            
                                                input_y -->   Reshape --->    
input_x -- >transpose
                                                     Const ----|            
                                                input_y -->   Reshape --->
</pre>
After fix:
Removes the dangling transpose.

### CODE REVIEW

File: onnxruntime/core/optimizer/constant_folding.cc [hunk 1]
```diff
@@ -174,6 +174,13 @@ Status ConstantFolding::ApplyImpl(Graph& graph, bool& modified, int graph_level,
     }
 
     if (converted_to_constant) {
+      // Remove single-output node chain for inputs of the node
+      for (int j = 0; j < node->InputArgCount().front(); ++j) {
```
Comment (SherlockNoMad) [inline onnxruntime/core/optimizer/constant_folding.cc:178]: for (auto iter = node->InputNodesBegin(); iter != node->InputNodesEnd(); ++iter)

File: onnxruntime/core/optimizer/concat_slice_elimination.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+@Class ConcatSliceElimination
+*/
+class ConcatSliceElimination : public GraphTransformer {
+ public:
+  ConcatSliceElimination(const std::unordered_set<std::string>& compatible_execution_providers = {}) noexcept
+      : GraphTransformer("ConcatSliceElimination", compatible_execution_providers) {}
+
+  Status ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const override;
+private:
+  static bool Fuse_Subgraph(Node& concat, Graph& graph, const logging::Logger& logger);
```
Comment (SherlockNoMad) [inline onnxruntime/core/optimizer/concat_slice_elimination.h:20]: >Fuse_Subgraph [](start = 14, length = 13)

Nit FuseSubgraph. or make it more explicit FuseConcatSliceSubgraph

File: onnxruntime/core/optimizer/concat_slice_elimination.cc [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+  std::vector<onnxruntime::Node*> node_consumers = graph.GetMutableConsumerNodes(concat.MutableOutputDefs()[0]->Name());
+  if (node_consumers.size() != 3) return false;
+
+  std::vector<onnxruntime::Node*> ordered_slice = node_consumers;
+  for (auto slice : node_consumers) {
+    std::vector<int64_t> starts, ends, axes, steps;
+    bool success = GetSliceInfo(graph, *slice, logger, starts, ends, axes, steps);
+    if (!success) return false;
+    if (starts.size() > 1) return false;
+    if (axes[0] != 0) return false;
```
Comment (SherlockNoMad) [inline onnxruntime/core/optimizer/concat_slice_elimination.cc:201]: axes[0] should match the concat's axis setting, right?

File: onnxruntime/core/optimizer/concat_slice_elimination.cc [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    if (steps[0] != 1) return false;
+    if (starts[0] == 0 && ends[0] == q_bias_len) {
+      ordered_slice[0] = slice;
+    } else if (starts[0] == q_bias_len && ends[0] == q_bias_len + k_bias_len) {
+      ordered_slice[1] = slice;
+    } else if (starts[0] == q_bias_len + k_bias_len) {
+      ordered_slice[2] = slice;
+    } else {
+      return false;
+    }
```
Comment (SherlockNoMad) [inline onnxruntime/core/optimizer/concat_slice_elimination.cc:211]: The way it's written here is not very ideal. 
maybe use two counters: start and end, and check them if they matches starts[i] and ends[i] ?

File: onnxruntime/core/optimizer/concat_slice_elimination.cc [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+  }
+  int replace_cnt = 0;
+  for (auto slice_node : ordered_slice) {
+    Node& add_node = *graph.GetNode(slice_node->OutputNodesBegin()->Index());
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(add_node, "Add", {7})) {
+      return false;
+    }
```
Comment (SherlockNoMad) [inline onnxruntime/core/optimizer/concat_slice_elimination.cc:218]: Why is this check needed? we don't care about the downstream nodes, right? 
I think we can relax this condition....

File: onnxruntime/core/optimizer/concat_slice_elimination.cc [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    graph_utils::RemoveNodeOutputEdges(graph, *slice_node);
+    graph_utils::ReplaceNodeInput(add_node, 1, *(concat_inputs[replace_cnt]));
+    replace_cnt++;
+  }
+  if (replace_cnt == 3) {
+    //delete the slice nodes and concat node
+    graph_utils::RemoveNodeOutputEdges(graph, concat);
+    graph.RemoveNode(concat.Index());
+    for (auto slice_node : ordered_slice) {
```
Comment (SherlockNoMad) [inline onnxruntime/core/optimizer/concat_slice_elimination.cc:227]: >       [](start = 43, length = 6)

white space

File: onnxruntime/core/optimizer/concat_slice_elimination.cc [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+v_bias              -->Slice [q+k, :]---> Add_v
+
+After transform:
+q_bias ---> Add_q  
+      
+k_bias ---> Add_k
+
+v_bias ---> Add_v
+
+*/
```
Comment (SherlockNoMad) [inline onnxruntime/core/optimizer/concat_slice_elimination.cc:163]: Overall, i think the pattern here is too specific for QKV... 
The same logic can be applied to Concat with >1 inputs...

File: onnxruntime/test/optimizer/graph_transform_test.cc [hunk 8]
```diff
@@ -1648,6 +1649,24 @@ TEST_F(GraphTransformationTests, ReshapeFusionDistilBertTest) {
...
+  std::shared_ptr<Model> p_model;
+  ASSERT_STATUS_OK(Model::Load(model_uri, p_model, nullptr, *logger_));
+  Graph& graph = p_model->MainGraph();
+
+  onnxruntime::GraphTransformerManager graph_transformation_mgr{5};
+  graph_transformation_mgr.Register(onnxruntime::make_unique<ConcatSliceElimination>(), TransformerLevel::Level1);
+  auto ret = graph_transformation_mgr.ApplyTransformers(graph, TransformerLevel::Level1, *logger_);
+  ASSERT_TRUE(ret.IsOK());
+
+  Model::Save(*p_model, "concat_slice_transformed.onnx");
```
Comment (SherlockNoMad) [inline onnxruntime/test/optimizer/graph_transform_test.cc:1664]: is the model save still needed? this was for debugging purpose right?

File: onnxruntime/core/optimizer/constant_folding.cc [hunk 9]
```diff
@@ -174,6 +174,10 @@ Status ConstantFolding::ApplyImpl(Graph& graph, bool& modified, int graph_level,
     }
 
     if (converted_to_constant) {
+      // Remove single-output node chain for inputs of the node
+      for (auto p_ip_node = node->InputNodesBegin(); p_ip_node != node->InputNodesEnd(); ++p_ip_node) {
+        graph_utils::RemoveNodesWithOneOutputBottomUp(graph, *p_ip_node);
```
Comment (SherlockNoMad) [inline onnxruntime/core/optimizer/constant_folding.cc:179]: I wonder how this works... 
From the code, it looks like it's deleting the input nodes of current node. 
How would it trace up to the chain and delete them in a row? 
Did I misunderstand something here?

Comment (SherlockNoMad) [reply onnxruntime/core/optimizer/constant_folding.cc:179]: synced offline. 
RemoveNodesWithOneOutputBottomUp takes care of the node removal in a chain

**Review by SherlockNoMad — DISMISSED** [review]

File: onnxruntime/core/optimizer/constant_folding.cc (hunk 9 above)
Comment (ashbhandare) [reply onnxruntime/core/optimizer/constant_folding.cc:179]: the function graph_utils::RemoveNodesWithOneOutputBottomUp() takes care of deleting the chain of unused inputs

File: onnxruntime/test/optimizer/graph_transform_test.cc (hunk 8 above)
Comment (ashbhandare) [reply onnxruntime/test/optimizer/graph_transform_test.cc]: that's right, removing

**Review by SherlockNoMad — APPROVED** [review]

**Pull request merged**